include/jwt-cpp/jwt.h

```diff
@@ -3267,219 +3705,34 @@ namespace jwt {
 			const typename json_traits::string_type data = jwt.get_header_base64() + "." + jwt.get_payload_base64();
 			const typename json_traits::string_type sig = jwt.get_signature();
 			const std::string algo = jwt.get_algorithm();
-			if (algs.count(algo) == 0) {
-				ec = error::token_verification_error::wrong_algorithm;
-				return;
+			std::string kid("");
+			if (jwt.has_header_claim("kid")) { kid = jwt.get_header_claim("kid").as_string(); }
+
+			typename keysets::const_iterator key_set_it = keys.find(kid);
+			bool key_found = false;
+			if (key_set_it != keys.end()) {
+				const key_list& keys = key_set_it->second;
+				for (const auto& key : keys) {
+					if (key.supports(algo)) {
+						key_found = true;
+						auto alg = from_key_and_alg(key, algo, ec);
+						alg->verify(data, sig, ec);
+						break;
+					}
+				}
 			}
-			algs.at(algo)->verify(data, sig, ec);
-			if (ec) return;
 
-			verify_ops::verify_context<json_traits> ctx{clock.now(), jwt, default_leeway};
-			for (auto& c : claims) {
-				ctx.claim_key = c.first;
-				c.second(ctx, ec);
-				if (ec) return;
+			if (!key_found) {
+				if (algs.count(algo) == 0) {
+					ec = error::token_verification_error::wrong_algorithm;
+					return;
+				}
+				algs.at(algo)->verify(data, sig, ec);
```

Comment:
I know this is my own code, but since we are already at changing it, this would be better:
```cpp
auto alg = algs.find(algo);
if(alg == algs.end()) {
      ec = error::token_verification_error::wrong_algorithm;
      return;
}
alg->second->verify(data, sig, ec);
```
Prevents the duplicate lookup.